contrib/endpoints/src/api_manager/config_manager.cc

```diff
@@ -0,0 +1,162 @@
+  if (global_context_->service_name().empty()) {
+    std::string msg = "API service name is not specified";
+    global_context_->env()->LogError(msg);
+    rollout_apply_function_(utils::Status(Code::ABORTED, msg), kEmptyConfigs);
+    return;
+  }
+
+  // TODO(jaebong) config_id should not be empty for the first version
+  // This part will be removed after the rollouts feature added
+  if (global_context_->config_id().empty()) {
+    std::string msg = "API config_id is not specified";
+    global_context_->env()->LogError(msg);
+    rollout_apply_function_(utils::Status(Code::ABORTED, msg), kEmptyConfigs);
+    return;
+  }
+
+  // Fetch service account token
+  GlobalFetchServiceAccountToken(global_context_, [this](utils::Status status) {
+    OnFetchAuthTokenDone(status);
+  });
+}
+
+void ConfigManager::OnFetchAuthTokenDone(utils::Status status) {
+  if (!status.ok()) {
+    // We should not get here
+    global_context_->env()->LogError("Unexpected status: " + status.ToString());
+    rollout_apply_function_(utils::Status(Code::ABORTED, status.ToString()),
+                            kEmptyConfigs);
+  }
+
+  // Fetch configs from the Inception
+  auto config_fetch_info = std::make_shared<ConfigsFetchInfo>();
+
+  // For now, config manager has only one config_id (100% rollout)
+  config_fetch_info->rollouts.assign({{global_context_->config_id(), 100}});
+  FetchConfigs(config_fetch_info);
+}
+
+// Fetch configs from rollouts. fetch_info has rollouts and fetched configs
+void ConfigManager::FetchConfigs(
+    std::shared_ptr<ConfigsFetchInfo> config_fetch_info) {
+  // Finished fetching configs.
+  if (config_fetch_info->IsCompleted()) {
+    // Failed to fetch all configs or rollouts are empty
+    if (config_fetch_info->IsRolloutsEmpty() ||
+        config_fetch_info->IsConfigsEmpty()) {
+      // first time, call the ApiManager callback function with an error
+      rollout_apply_function_(
+          utils::Status(Code::ABORTED, "Failed to download the service config"),
+          kEmptyConfigs);
+      return;
+    }
+
+    // Update ApiManager
+    rollout_apply_function_(utils::Status::OK, config_fetch_info->configs);
+    return;
+  }
+
+  service_management_fetch_->GetConfig(
+      config_fetch_info->rollouts[config_fetch_info->index].first,
+      [this, config_fetch_info](const utils::Status& status,
+                                const std::string& config) {
+        if (status.ok()) {
+          config_fetch_info->configs.push_back(
+              {std::move(config),
+               config_fetch_info->rollouts[config_fetch_info->index].second});
+        } else {
+          global_context_->env()->LogError(std::string(
+              "Unable to download Service config for the config_id: " +
+              config_fetch_info->rollouts[config_fetch_info->index].first));
+        }
+
+        // move on to the next config_id
+        config_fetch_info->Next();
```

Comment:
why do we want to fetch configs sequentially?  Can we try to fetch all configs in parallel?